src/controller/python/chip/clusters/command.cpp

```diff
@@ -181,6 +237,83 @@ PyChipError pychip_CommandSender_SendCommand(void * appContext, DeviceProxy * de
     return ToPyChipError(err);
 }
 
+PyChipError pychip_CommandSender_SendBatchCommands(void * appContext, DeviceProxy * device, uint16_t timedRequestTimeoutMs,
+                                                   uint16_t interactionTimeoutMs, uint16_t busyWaitMs, bool suppressResponse,
+                                                   size_t n, ...)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    VerifyOrReturnError(device->GetSecureSession().HasValue(), ToPyChipError(CHIP_ERROR_MISSING_SECURE_SESSION));
+    auto remoteSessionParameters = device->GetSecureSession().Value()->GetRemoteSessionParameters();
+    CommandSender::ConfigParameters config;
+    config.SetRemoteMaxPathsPerInvoke(remoteSessionParameters.GetMaxPathsPerInvoke());
+
+    std::unique_ptr<CommandSenderCallback> callback =
+        std::make_unique<CommandSenderCallback>(appContext, /* isBatchedCommands =*/true);
+    std::unique_ptr<CommandSender> sender =
+        std::make_unique<CommandSender>(callback.get(), device->GetExchangeManager(),
+                                        /* is timed request */ timedRequestTimeoutMs != 0, suppressResponse);
+
+    va_list args;
+    va_start(args, n);
+
+    SuccessOrExit(err = sender->SetCommandSenderConfig(config));
+
+    {
+        for (size_t i = 0; i < n; i++)
```

Comment:
Seems like this whole block should be somewhere in InvokeInteraction - this looks correct, just seems like an odd place for it since this looks like fairly general c++ code sitting in the python dir.

Reply:
yeah, but isn't the TLV generation something that all batch command senders need to do?

Reply:
If you are referring to this part specifically:
```

            SuccessOrExit(err = sender->PrepareCommand(cmdParams, additionalParams));
            {
                auto writer = sender->GetCommandDataIBTLVWriter();
                VerifyOrExit(writer != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
                TLV::TLVReader reader;
                reader.Init(tlvBuffer, static_cast<uint32_t>(length));
                reader.Next();
                SuccessOrExit(err = writer->CopyContainer(TLV::ContextTag(CommandDataIB::Tag::kFields), reader));
            }

            SuccessOrExit(err = sender->FinishCommand(timedRequestTimeoutMs != 0 ? Optional<uint16_t>(timedRequestTimeoutMs)
                                                                                 : Optional<uint16_t>::Missing(),
                                                      additionalParams));
```

Then yes `CommandSender` does have a helper, and most things are calling `AddRequestData`, but that is a templated function that takes in something that call the Encode into TLV. When you have the raw TLV this is how it is currently done. I guess we could provide a helper function for that as well. But thinking one step down the line for chip-repl for automated testing we are going to want to inject non-identical `CommandRef` a so we will need to be able to do it this particular way anyways